File: src/main/java/com/regionrental/config/StorageConfig.java
package com.regionrental.config;

import com.regionrental.RegionRental;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.inventory.ItemStack;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.logging.Level;

public class StorageConfig {
    
    private final RegionRental plugin;
    private File configFile;
    private FileConfiguration config;
    
    public StorageConfig(RegionRental plugin) {
        this.plugin = plugin;
        createConfig();
    }
    
    private void createConfig() {
        configFile = new File(plugin.getDataFolder(), "storage.yml");
        
        if (!configFile.exists()) {
            configFile.getParentFile().mkdirs();
            try {
                configFile.createNewFile();
                plugin.getLogger().info("Created storage.yml file");
            } catch (IOException e) {
                plugin.getLogger().log(Level.SEVERE, "Could not create storage.yml!", e);
            }
        }
        
        config = YamlConfiguration.loadConfiguration(configFile);
    }
    
    public void save() {
        try {
            config.save(configFile);
        } catch (IOException e) {
            plugin.getLogger().log(Level.SEVERE, "Could not save storage.yml!", e);
        }
    }
    
    public void reload() {
        config = YamlConfiguration.loadConfiguration(configFile);
    }
    
    public void storeItems(UUID playerUUID, String region, List<ItemStack> items) {
        storeItems(playerUUID, region, items, new ArrayList<>());
    }

    public void storeItems(UUID playerUUID, String region, List<ItemStack> containerItems, List<ItemStack> blockItems) {
        String path = "storage." + playerUUID.toString() + "." + region;
        config.set(path + ".timestamp", System.currentTimeMillis());
        config.set(path + ".items", containerItems);
        config.set(path + ".blocks", blockItems);
        save();
    }
    
    public List<ItemStack> retrieveItems(UUID playerUUID, String region) {
        String path = "storage." + playerUUID.toString() + "." + region + ".items";
        
        if (!config.contains(path)) {
            return new ArrayList<>();
        }
        
        @SuppressWarnings("unchecked")
        List<ItemStack> items = (List<ItemStack>) config.getList(path, new ArrayList<>());
        
        // Remove the items after retrieval
        config.set("storage." + playerUUID.toString() + "." + region, null);
        
        // Clean up empty sections
        if (config.getConfigurationSection("storage." + playerUUID.toString()) != null &&
            config.getConfigurationSection("storage." + playerUUID.toString()).getKeys(false).isEmpty()) {
            config.set("storage." + playerUUID.toString(), null);
        }
        
        save();
        return items;
    }
    
    public List<ItemStack> getAllStoredItems(UUID playerUUID) {
        List<ItemStack> allItems = new ArrayList<>();
        String basePath = "storage." + playerUUID.toString();

        if (!config.contains(basePath)) {
            return allItems;
        }

        for (String region : config.getConfigurationSection(basePath).getKeys(false)) {
            // Get container items
            @SuppressWarnings("unchecked")
            List<ItemStack> items = (List<ItemStack>) config.getList(basePath + "." + region + ".items", new ArrayList<>());
            allItems.addAll(items);

            // Get block items
            @SuppressWarnings("unchecked")
            List<ItemStack> blocks = (List<ItemStack>) config.getList(basePath + "." + region + ".blocks", new ArrayList<>());
            allItems.addAll(blocks);
        }

        return allItems;
    }
    
    public void clearPlayerStorage(UUID playerUUID) {
        config.set("storage." + playerUUID.toString(), null);
        save();
    }

    /**
     * Updates storage with remaining items after partial retrieval
     * If no items remain, clears storage completely
     * @param playerUUID The player's UUID
     * @param remainingItems List of items that weren't taken from the GUI
     */
    public void updatePartialStorage(UUID playerUUID, List<ItemStack> remainingItems) {
        if (remainingItems.isEmpty()) {
            // No items left, clear completely
            clearPlayerStorage(playerUUID);
            return;
        }

        // Clear existing storage first
        clearPlayerStorage(playerUUID);

        // Re-store remaining items under a generic region name
        // Note: Original region information is lost, but players just want their items back
        storeItems(playerUUID, "partial_retrieval", remainingItems, new ArrayList<>());
    }

    public boolean hasStoredItems(UUID playerUUID) {
        return config.contains("storage." + playerUUID.toString());
    }
    
    public List<String> getStoredRegions(UUID playerUUID) {
        List<String> regions = new ArrayList<>();
        String basePath = "storage." + playerUUID.toString();
        
        if (!config.contains(basePath)) {
            return regions;
        }
        
        regions.addAll(config.getConfigurationSection(basePath).getKeys(false));
        return regions;
    }
    
    public long getStorageTimestamp(UUID playerUUID, String region) {
        String path = "storage." + playerUUID.toString() + "." + region + ".timestamp";
        return config.getLong(path, 0L);
    }
    
    public void cleanupOldStorage(long maxAge) {
        long currentTime = System.currentTimeMillis();
        
        if (!config.contains("storage")) {
            return;
        }
        
        for (String uuid : config.getConfigurationSection("storage").getKeys(false)) {
            for (String region : config.getConfigurationSection("storage." + uuid).getKeys(false)) {
                long timestamp = config.getLong("storage." + uuid + "." + region + ".timestamp", 0L);
                
                if (currentTime - timestamp > maxAge) {
                    config.set("storage." + uuid + "." + region, null);
                    plugin.getLogger().info("Cleaned up old storage for " + uuid + " in region " + region);
                }
            }
        }
        
        save();
    }
}

File: src/main/kotlin/com/regionrental/config/StorageConfig.kt
package com.regionrental.config

import com.regionrental.RegionRental
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.inventory.ItemStack
import java.io.File
import java.io.IOException
import java.util.UUID
import java.util.logging.Level

class StorageConfig(private val plugin: RegionRental) {

    private val configFile: File = File(plugin.dataFolder, "storage.yml")
    private var config: YamlConfiguration

    @Volatile
    var isDirty: Boolean = false
        private set

    init {
        createConfig()
        config = YamlConfiguration.loadConfiguration(configFile)
    }

    private fun createConfig() {
        if (!configFile.exists()) {
            configFile.parentFile?.mkdirs()
            try {
                configFile.createNewFile()
                plugin.logger.info("Created storage.yml file")
            } catch (e: IOException) {
                plugin.logger.log(Level.SEVERE, "Could not create storage.yml!", e)
            }
        }
    }

    private fun markDirty() {
        isDirty = true
    }

    fun saveIfDirty() {
        if (isDirty) save()
    }

    fun save() {
        try {
            config.save(configFile)
        } catch (e: IOException) {
            plugin.logger.log(Level.SEVERE, "Could not save storage.yml!", e)
            return
        }
        isDirty = false
    }

    fun reload() {
        config = YamlConfiguration.loadConfiguration(configFile)
        isDirty = false
    }

    fun storeItems(playerUUID: UUID, region: String, items: List<ItemStack>) {
        storeItems(playerUUID, region, items, emptyList())
    }

    fun storeItems(playerUUID: UUID, region: String, containerItems: List<ItemStack>, blockItems: List<ItemStack>) {
        val path = "storage.$playerUUID.$region"
        config.set("$path.timestamp", System.currentTimeMillis())
        config.set("$path.items", containerItems)
        config.set("$path.blocks", blockItems)
        markDirty()
    }

    fun retrieveItems(playerUUID: UUID, region: String): List<ItemStack> {
        val path = "storage.$playerUUID.$region.items"

        if (!config.contains(path)) {
            return mutableListOf()
        }

        @Suppress("UNCHECKED_CAST")
        val items = config.getList(path, mutableListOf<ItemStack>()) as? List<ItemStack> ?: mutableListOf()

        // Remove the items after retrieval
        config.set("storage.$playerUUID.$region", null)

        // Clean up empty sections
        config.getConfigurationSection("storage.$playerUUID")?.let { section ->
            if (section.getKeys(false).isEmpty()) {
                config.set("storage.$playerUUID", null)
            }
        }

        markDirty()
        return items
    }

    fun getAllStoredItems(playerUUID: UUID): List<ItemStack> {
        val allItems = mutableListOf<ItemStack>()
        val basePath = "storage.$playerUUID"

        val section = config.getConfigurationSection(basePath) ?: return allItems

        for (region in section.getKeys(false)) {
            @Suppress("UNCHECKED_CAST")
            val items = config.getList("$basePath.$region.items", mutableListOf<ItemStack>()) as? List<ItemStack>
            items?.let { allItems.addAll(it) }

            @Suppress("UNCHECKED_CAST")
            val blocks = config.getList("$basePath.$region.blocks", mutableListOf<ItemStack>()) as? List<ItemStack>
            blocks?.let { allItems.addAll(it) }
        }

        return allItems
    }

    fun clearPlayerStorage(playerUUID: UUID) {
        config.set("storage.$playerUUID", null)
        markDirty()
    }

    /**
     * Updates storage with remaining items after partial retrieval
     * If no items remain, clears storage completely
     */
    fun updatePartialStorage(playerUUID: UUID, remainingItems: List<ItemStack>) {
        if (remainingItems.isEmpty()) {
            clearPlayerStorage(playerUUID)
            return
        }

        // Clear existing storage first
        clearPlayerStorage(playerUUID)

        // Re-store remaining items under a generic region name
        storeItems(playerUUID, "partial_retrieval", remainingItems, emptyList())
    }

    fun hasStoredItems(playerUUID: UUID): Boolean =
        config.contains("storage.$playerUUID")

    fun getStoredRegions(playerUUID: UUID): List<String> {
        val basePath = "storage.$playerUUID"
        val section = config.getConfigurationSection(basePath) ?: return emptyList()
        return section.getKeys(false).toList()
    }

    fun getStorageTimestamp(playerUUID: UUID, region: String): Long =
        config.getLong("storage.$playerUUID.$region.timestamp", 0L)

    fun cleanupOldStorage(maxAge: Long) {
        val currentTime = System.currentTimeMillis()

        val storageSection = config.getConfigurationSection("storage") ?: return

        for (uuid in storageSection.getKeys(false)) {
            val playerSection = config.getConfigurationSection("storage.$uuid") ?: continue

            for (region in playerSection.getKeys(false)) {
                val timestamp = config.getLong("storage.$uuid.$region.timestamp", 0L)

                if (currentTime - timestamp > maxAge) {
                    config.set("storage.$uuid.$region", null)
                    plugin.logger.info("Cleaned up old storage for $uuid in region $region")
                }
            }
        }

        markDirty()
    }
}
